### PR TITLE
Fix rolling update of existing pods when scaling up/down

### PR DESCRIPTION
**Pull request by scholzj**
### Type of change

- Bugfix

### Description

When checking the update of the secrets with server certificates for our different STS / Deployments, we trigger unnecessary rolling update when scaling up or down (new records are added to the secret -> the secret changed and we trigger rolling update of all pods.). This PR checks the changes in more details so that it triggers the rolling update only when the certificates which eisted before changed. Thanks to that Kafka doesn't roll when it scales up or down.

### Checklist

- [x] Make sure all tests pass
- [x] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally

### CODE REVIEW

**Comment by scholzj** [comment]
@ppatierno @tombentley I do not think this necessarily needs review from both of you ... but at least one would be good ;-)

**Comment by tombentley** [comment]
@ppatierno I'll take this.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/ModelUtils.java [hunk 1]
```diff
@@ -366,4 +366,31 @@ public static String encodeStorageToJson(Storage storage) {
...
+     * changed. This method is used to evaluate whether rolling update of existing brokers is needed when secrets with
+     * certificates change. It separates changes for existing certificates with other changes to the secret such as
+     * added or removed certificates (scale-up or scale-down).
+     *
+     * @param current   Existing secret
+     * @param desired   Desired secret
+     *
+     * @return  True if there is a key which exists in the data sections of both secrets and which changed.
+     */
+    public static boolean didAnyCertificateChangedInSecret(Secret current, Secret desired) {
```
Comment (tombentley) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/ModelUtils.java]: I wonder if the name `(do)commonCertificatesDiffer` is better?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java [hunk 2]
```diff
@@ -2866,12 +2884,18 @@ String getInternalServiceHostname(String serviceName)    {
...
+                            .map(res -> {
+                                if (res instanceof ReconcileResult.Patched) {
+                                    // The secret is patched and some changes to the existing certificates actually occured
+                                    isExistingKafkaExporterCertsChanged = ModelUtils.didAnyCertificateChangedInSecret(oldSecret, res.resource());
+                                }
+
+                                return this;
+                            })
+                    );
         }
```
Comment (tombentley) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java]: We've got this same pattern of method calls repeated three or four times in this PR. Surely we can factor out a common method which takes the secret name and `BiFunction` whose argument is a method reference to `kafkaExporter::generateSecret` or whatever?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java [hunk 3]
```diff
@@ -383,10 +383,10 @@ ReconciliationState createReconciliationState(Reconciliation reconciliation, Kaf
...
 
         // Certificate change indicators
-        private boolean isZookeeperCertsChanged = false;
-        private boolean isKafkaCertsChanged = false;
-        private boolean isKafkaExporterCertsChanged = false;
-        private boolean isEntityOperatorCertsChanged = false;
+        private boolean isExistingZookeeperCertsChanged = false;
+        private boolean isExistingKafkaCertsChanged = false;
+        private boolean isExistingKafkaExporterCertsChanged = false;
+        private boolean isExistingEntityOperatorCertsChanged = false;
```
Comment (tombentley) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java]: If we're changing the name of these anyway, I wonder if we should drop the `is` part. The tense disagrees with the `changed` suffix, so it reads a bit awkwardly.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/ModelUtils.java (hunk 1 above)
Comment (scholzj) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/model/ModelUtils.java]: I think `common` is a bit confusing. I went with `doExistingCertificatesDiffer` - hope that works for you.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java (hunk 2 above)
Comment (scholzj) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java]: I migth have done this a bit differently, but I factored it out. and re-used it in all 4 methods.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java (hunk 3 above)
Comment (scholzj) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java]: Yeah, now when you say it it sounds weird :-D

**Comment by scholzj** [comment]
@strimzi-ci run tests

**Comment by scholzj** [comment]
@strimzi-ci run tests profile=acceptance testcase=io.strimzi.systemtest.RollingUpdateST#testKafkaAndZookeeperScaleUpScaleDown

**Pull request merged**